### ISSUE TITLE
Invoke-DbaQuery Exception The specified module could not be found Microsoft.Data.SqlClient.SNI.x64.dll

### ISSUE THREAD

**Issue by szeltner-wp** [dataplat/dbatools#9015]
### Verified issue does not already exist?

I have searched and found no existing issue

### What error did you receive?

Error connecting to [XXXXX]: Unable to load DLL 'Microsoft.Data.SqlClient.SNI.x64.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)

### Steps to Reproduce

```powershell
# provide your command(s) executed pertaining to dbatools
# please include variable values (redacted or fake if needed) for reference
```
Invoke-DbaQuery -SqlInstance $DBServer -Query $Query -Database $DB -SqlCredential $cred -EnableException -As SingleValue

### Please confirm that you are running the most recent version of dbatools

2.0.4

### Other details or mentions

Registering runspace: dbatools-logging
Starting runspace: dbatools-logging
Registering runspace: dbatools-teppasynccache
Starting runspace: dbatools-teppasynccache
Registering runspace: dbatools-maintenance
Starting runspace: dbatools-maintenance
Updating runspace: dbatools-logging
Starting runspace: dbatools-logging
Updating runspace: dbatools-teppasynccache
Starting runspace: dbatools-teppasynccache
Updating runspace: dbatools-maintenance
Starting runspace: dbatools-maintenance
Updating runspace: dbatools-logging
Starting runspace: dbatools-logging
Updating runspace: dbatools-teppasynccache
Starting runspace: dbatools-teppasynccache
Updating runspace: dbatools-maintenance
Starting runspace: dbatools-maintenance
Bound parameters: SqlInstance, Query, Database, SqlCredential, EnableException, As
SqlInstance passed in, will work on: XXXXXX
Starting process block
Immediately checking for Azure
Test for Azure is negative
Starting loop for 'XXXXX': ComputerName = 'XXXXXX', InstanceName = 'XXXXXX', IsLocalHost = 'False', Type = 'Default'
String is passed in, will build server object from instance object and other parameters, do some checks and then return the server object
Test for Azure is negative
authentication method is 'local sql'
ApplicationName will be set to 'dbatools PowerShell module - dbatools.io'
ConnectionTimeout will be set to '15'
Database will be set to 'XXXXXX'
EncryptConnection will be set to 'False'
PacketSize will be set to '4096'
Pooled will be set to 'False'
SecurePassword will be set
TrustServerCertificate will be set to 'True'
UserName will be set to 'XXXXXX'
Building ServerConnection from SqlConnectionInfo
ServerConnection was built
Building Server from ServerConnection
Server was built
Setting ConnectionContext.StatementTimeout to '0'
The masked server.ConnectionContext.ConnectionString is Data Source=XXXXXX;Initial Catalog=XXXXXX;User ID=XXXXXX;Password=********;Pooling=False;Multiple Active Result
Sets=False;Encrypt=True;Trust Server Certificate=True;Packet Size=4096;Application Name="dbatools PowerShell module - dbatools.io"
We connect to the instance by running SELECT 'dbatools is opening a new connection'
Failure | Error connecting to [XXXXXX]: Unable to load DLL 'Microsoft.Data.SqlClient.SNI.x64.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)
Failure | Unable to load DLL 'Microsoft.Data.SqlClient.SNI.x64.dll': The specified module could not be found. (Exception from HRESULT: 0x8007007E)

When I do this in x86 it throws "Microsoft.Data.SqlClient.SNI.x86.dll"
When I do this in PS7 it throws InvalidOperation:
  Line |
151381 |          throw $records[0]
       |          ~~~~~~~~~~~~~~~~~
       | Could not load type 'Microsoft.SqlServer.Server.SqlContext' from assembly 'System.Data, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089'.

I've also tried in vscode terminal with the same results

### What PowerShell host was used when producing this error

PowerShell Core (pwsh.exe), Windows PowerShell (powershell.exe), VS Code (terminal), VS Code (integrated terminal)

### PowerShell Host Version

Name                           Value
----                           -----
PSVersion                      7.3.4
PSEdition                      Core
GitCommitId                    7.3.4
OS                             Microsoft Windows 10.0.17763
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

Name                           Value                                                                                   
----                           -----                                                                                   
PSVersion                      5.1.17763.3770                                                                          
PSEdition                      Desktop                                                                                 
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                 
BuildVersion                   10.0.17763.3770                                                                         
CLRVersion                     4.0.30319.42000                                                                         
WSManStackVersion              3.0                                                                                     
PSRemotingProtocolVersion      2.3                                                                                     
SerializationVersion           1.1.0.1                                                                                 

### SQL Server Edition and Build number

Microsoft SQL Server 2016 (SP2-CU5) (KB4475776) - 13.0.5264.1 (X64)  	Jan 10 2019 18:51:38  	Copyright (c) Microsoft Corporation 	Enterprise Edition: Core-based Licensing (64-bit) on Windows Server 2012 R2 Standard 6.3 <X64> (Build 9600: ) (Hypervisor) 

### .NET Framework Version

PSChildName                      Version
-----------                      -------
v2.0.50727                       2.0.50727.4927
v3.0                             3.0.30729.4926
Windows Communication Foundation 3.0.4506.4926
Windows Presentation Foundation  3.0.6920.4902
v3.5                             3.5.30729.4926
Client                           4.8.03761
Full                             4.8.03761
Client                           4.0.0.0

**Comment by andreasjordan** [comment]
Can you validate that the module dbatools.library is installed correctly? The dll file should be in "C:\Program Files\WindowsPowerShell\Modules\dbatools.library\2023.5.5\desktop\lib".

And can you check that the module SqlServer is NOT loaded in that session?

**Comment by andreasjordan** [comment]
Maybe duplicate of #8268

**Comment by andreasjordan** [comment]
What kind of Credential do you use as `$cred` in `-SqlCredential $cred`?

And if it's a windows user - can you start the powershell session as that user and then use integrated security?

**Comment by szeltner-wp** [comment]
Sorry for the delay and thank you for the reply.
This was an environmental issue and not a problem with DBA Tools.  Allow me to explain:
This was a situation where I was running 64-bit Powershell 5.1 code from an Amazon workspace (server 2019) through a C# .NET 4.6 Runspace (System.Management.Automation.dll) through an Automation Platform (Blue Prism).  In other words, pretty tricky to debug due to dll sandboxing, security permissions etc.

What I didn't realize was that my "C:\" drive must be network mount of some sort in the Amazon Workspace.  No matter what I did, I was not able to get dbatools.library to load due to HRESULT "Operation Not Supported" on seemingly random dlls (all the dlls in the "third party" folder and sometimes Microsoft.Identity or Microsoft.Data.***)  In my research, most of these issues were a result of needing to unblock the DLL.  The dlls never showed as blocked.

Why didn't I move everything to "local" storage?  I thought I did but this workspace must not have "local" storage in the sense that Windows thinks that it is local.

So a thousand red herrings later, I found that I needed to edit the C:\Windows\Microsoft.NET\Framework64\v4.0.30319\Config\machine.config file and add
     <runtime>
         <loadFromRemoteSources enabled="true"/>
    </runtime>

I realize this may be a bit too broad in scope but doing it just to Powershell.exe.config didn't do the trick.


Procmon was helpful in this
Procexplorer was VERY helpful in finding what dlls were loaded.

Closing the issue.  Thank you!
